internal/sys/fd_trace.go

```diff
@@ -0,0 +1,137 @@
+package sys
+
+import (
+	"bytes"
+	"fmt"
+	"runtime"
+	"strings"
+	"sync"
+)
+
+// FDs is a registry of all file descriptors wrapped into sys.FDs that were
+// created while an fd tracer was active.
+var FDs = newFDMap()
+
+// Finalize is called when the Go runtime frees an [FD].
+var Finalize FDTracer
+
+// traceFDs returns true if there are active fd tracers registered to the
+// package.
+func traceFDs() bool {
+	return Finalize != nil
+}
+
+type fdmap struct {
+	mu sync.RWMutex
+	m  map[int]FDTrace
+}
+
+func newFDMap() fdmap {
+	return fdmap{
+		m: make(map[int]FDTrace),
+	}
+}
+
+func (fdm *fdmap) Get(key int) (FDTrace, bool) {
+	fdm.mu.RLock()
+	defer fdm.mu.RUnlock()
+
+	v, ok := fdm.m[key]
+	if !ok {
+		return FDTrace{}, ok
+	}
+
+	return v, true
+}
+
+func (fdm *fdmap) Put(k int, value FDTrace) {
+	fdm.mu.Lock()
+	defer fdm.mu.Unlock()
+
+	fdm.m[k] = value
+}
+
+func (fdm *fdmap) Delete(k int) {
+	fdm.mu.Lock()
+	defer fdm.mu.Unlock()
+
+	delete(fdm.m, k)
+}
+
+func (fdm *fdmap) Len() int {
+	fdm.mu.RLock()
+	defer fdm.mu.RUnlock()
+
+	return len(fdm.m)
+}
+
+func (fdm *fdmap) String() string {
+	fdm.mu.RLock()
+	defer fdm.mu.RUnlock()
+
+	var b strings.Builder
+	for _, trace := range fdm.m {
+		b.WriteString(trace.String())
+		b.WriteString("\n")
+	}
+
+	return b.String()
+}
+
+// FDTrace is an event passed to an [FDTracer].
+type FDTrace struct {
+	// The file descriptor assigned to the resource.
+	FD int
+
+	// Stack trace at the point the resource was created.
+	Stack *runtime.Frames
+}
+
+func (t FDTrace) String() string {
+	return fmt.Sprintf("fd %d created at:\n%s", t.FD, formatFrames(t.Stack))
+}
+
+// An FDTracer is a function receiving an [FDTrace].
+// Assign it to [Finalize], etc.
+type FDTracer func(FDTrace)
+
+func (f FDTracer) do(fd *FD) {
+	t, ok := FDs.Get(fd.Int())
+	if !ok {
+		return
+	}
+
+	f(t)
+}
+
+func callersFrames() *runtime.Frames {
+	c := make([]uintptr, 32)
+	for {
+		// Skip runtime.Callers and this function.
+		i := runtime.Callers(2, c)
```

Comment:
Like I said before, retrying doesn't seem like the right choice here. "32 frames should be enough for everyone" applies I think. We just need a good enough stack to be able to figure out where the object was created. All of this CallersFrames stuff is a bit black magic, and I'd prefer to stick as close to the examples as possible.

Reply:
Ack on hardcoding 32 frames, that's fine for internal API. I found the dynamic resizing somewhere on Sourcegraph, but not sure if it was from the stdlib. This should be fine to do either way, but I'll remove it.